neet/sensitivity.py

```diff
@@ -35,33 +39,178 @@ def sensitivity(net, state):
+    Returns matrix answering the question:
+    Starting at the given state, does flipping the state of node j
+    change the state of node i?
+    """
+    # set up empty matrix
+    N = len(state)
+    Q = np.empty((N,N))
+    
+    # list Hamming neighbors (in order!)
+    neighbors = _hamming_neighbors(state)
+
+    nextState = net.update(state)
+
+    # count differences found in neighbors of the original
+    s = 0.
+    for j,neighbor in enumerate(neighbors):
+        if transitions is not None:
+            newState = transitions[_fast_encode(neighbor)]
+        else:
+            newState = net.update(neighbor)
+        Q[:,j] = [ ( nextState[i] + newState[i] )%2 for i in range(N) ]
+        
+    return Q
+
+def _states_limited(nodes,state0):
+    """
+    All possible states that vary only nodes with given indices.
+    """
+    if len(nodes) == 0:
+        return [state0]
+    for i in nodes:
+        stateFlipped = copy.copy(state0)
+        stateFlipped[nodes[0]] = (stateFlipped[nodes[0]]+1)%2
+        return _states_limited(nodes[1:],state0) + _states_limited(nodes[1:],stateFlipped)
+
+def _connections(net,nodei):
+    return net.table[nodei][0]
+
+def average_difference_matrix(net,states=None,weights=None,calc_trans=True):
+    """
+    Averaged over states, what is the probability
+    that node i's state is changed by a single bit flip of node j?
+    
+    states (None)       : If None, average over all possible states. (For logic
+                          networks, this case uses an algorithm that makes use of
+                          sparse connectivity to be much more efficient.)  
+                          Otherwise, providing a list of states will calculate 
+                          the average over only those states.
+    calc_trans (True)   : Optionally pre-calculate all transitions.  Only used
+                          when states argument is not None.
+    """
+    N = net.size
+    Q = np.zeros((N,N))
+
+    if (states is not None) or (weights is not None) or (not hasattr(net,'table')):
+        # explicitly calculate difference matrix for each state
+    
+        # optionally pre-calculate transitions
+        if calc_trans:
+            trans = list(transitions(net))
+        else:
+            trans = None
+    
+        # currently changes state generators to lists.
+        # is there a way to avoid this?
+        if states is None:
+            states = list(net.state_space())
+        else:
+            states = list(states)
+
+        if weights is None:
+            weights = np.ones(len(states))
+        else:
+            weights = list(weights)
+
+        if np.shape(weights) != (len(states),):
+            raise(ValueError("Weights must be a 1D array with length same as states"))
+
+        norm = np.sum(weights)
+        for i,state in enumerate(states):
+            Q += weights[i] * difference_matrix(net, state, trans) / norm
+
+    else: # make use of sparse connectivity to be more efficient
```

Comment:
Two questions here:
1. Will this algorithm slow down if the graph is not sparse? If you run this on a complete graph, will it be slower than the original algorithm?
2. Is there a reason this algorithm would only work for `LogicNetwork` instances? I see that the implementation uses the `table` attribute to get the incoming edges of given node, but that seems to be the only detail that makes this specific to `LogicNetwork` instances.

Reply:
1. I haven't tested it, but I'm almost certain it will be slower for dense networks.  Maybe we eventually want some kind of check on the network density?
2. I think we should be able to do this for other networks with sparse dependence structure.  We'd just need to define `_connections` to return the nodes that node i depends on.  (Is there an easy way to do this for WTNetworks?)

Reply:
1. I'm sure there's some way of quantifying sparseness. Maybe a feature for the future.
2. For `WTNetworks`, the incoming nodes are the indices with non-zero entries in the row (or column, I never remember which) of the weights matrix. I am weighing the possibility of requiring something like this as part of the interface requirements of a network: an object isn't a network unless there's some notion of node connectivity. Thoughts?

Reply:
1. Agreed.
2. Yes, makes sense to me.  (e.g. a natural thing to want to be able to do for any network is draw its nodes and edges)